### PR TITLE
PR: Fix slow find in large files

### LINKED ISSUE #14176: Spyder crashes when replacing text
## Problem Description
I have the (poorly written) python script below. I want to replace all instances of "particles" with "self.particles". To do so, I do [Search] > [Replace text] and try to replace "particles" with "self.particles". Spyder crashes immediately. I was having the issue with Spyder 4.1.2 inside Anaconda. I installed Spyder 4.2.0 using the stand-alone installer, but the error still exists.

### What steps reproduce the problem?

1.  Copy the **EXACT** following code into an editor window without saving

```
import rhinoscriptsyntax as rs
import random

particles = []
speedmax = 10
cycles = 20
sepval =5
alignval = 0.02 
chval = 0.02

class Particles:
    
    
    
    def __init__(self):
        
        pass

    def run(self):
        
        pass


    def flock(self):
              
        self.separate(sepval)
        self.cohesion(chval)
        self.align(alignval)

    def separate(self, magnitude):
        count = 0
        steer = [0.001,0.001,0.001]
        
        for i in range(0, len(particles)):
             dist = rs.Distance(self.coord,particles[i].coord)
             if dist>0 and dist <30 :
                 diff = rs.VectorSubtract(self.coord,particles[i].coord)
                 diff = rs.VectorUnitize(diff)
                 diff = rs.VectorScale(diff,1/dist)
                 steer = rs.VectorAdd(steer,diff)
                 count = count + 1
        if count > 0:
            steer = rs.VectorScale(steer,1/count)
        
        steer = rs.VectorScale(steer,magnitude)
        self.acc = rs.VectorAdd(self.acc,steer)


for i in range(nopoints):
    particles.append(Particles())

for i in range(cycles):
    for particle in particles:
        #particle.drawPosition()
        particle.run()

#for particle in particles:
#    rs.AddCurve(particle.path)
```

2. Click "CTRL+R" and replace "particles" with "self.particles"
3. Click "ALL"

Note: I tried several times to reduce the code to nail down the error. The extremely weird thing is that it's enough to replace one method's content with `pass` or deleting one of the rows at the beginning of the file containing those variables to prevent the error to occur. 

However, with this EXACT content, I can reproduce the error systematically.

## Versions

* Spyder version:  4.2.0
* Python version: 3.7.9 64-bit
* Qt version: 5.12.9
* PyQt version: 5.12.3
* Operating System name/version: Windows 10 

### Dependencies

```

# Mandatory:
atomicwrites >=1.2.0            :  1.4.0 (OK)
chardet >=2.0.0                 :  3.0.4 (OK)
cloudpickle >=0.5.0             :  1.6.0 (OK)
diff_match_patch >=20181111     :  20200713 (OK)
intervaltree                    :  None (OK)
IPython >=4.0                   :  7.19.0 (OK)
jedi =0.17.2                    :  0.17.2 (OK)
jsonschema >=3.2.0              :  3.2.0 (OK)
nbconvert >=4.0                 :  6.0.7 (OK)
numpydoc >=0.6.0                :  1.1.0 (OK)
paramiko >=2.4.0                :  2.7.2 (OK)
parso =0.7.0                    :  0.7.0 (OK)
pexpect >=4.4.0                 :  4.8.0 (OK)
pickleshare >=0.4               :  0.7.5 (OK)
psutil >=5.3                    :  5.7.3 (OK)
pygments >=2.0                  :  2.7.2 (OK)
pylint >=1.0                    :  2.6.0 (OK)
pyls >=0.36.1;<1.0.0            :  0.36.1 (OK)
pyls_black >=0.4.6              :  0.4.6 (OK)
pyls_spyder >=0.1.1             :  0.1.1 (OK)
qdarkstyle >=2.8                :  2.8.1 (OK)
qtawesome >=0.5.7               :  1.0.1 (OK)
qtconsole >=4.7.7               :  4.7.7 (OK)
qtpy >=1.5.0                    :  1.9.0 (OK)
setuptools >=39.0.0             :  50.3.2 (OK)
sphinx >=0.6.6                  :  3.3.0 (OK)
spyder_kernels >=1.10.0;<1.11.0 :  1.10.0 (OK)
three_merge >=0.1.1             :  0.1.1 (OK)
watchdog                        :  None (OK)
zmq >=17                        :  19.0.2 (OK)

# Optional:
cython >=0.21                   :  0.29.21 (OK)
matplotlib >=2.0.0              :  3.3.2 (OK)
numpy >=1.7                     :  1.19.4 (OK)
pandas >=1.1.1                  :  1.1.4 (OK)
scipy >=0.17.0                  :  1.5.4 (OK)
sympy >=0.7.3                   :  1.6.2 (OK)

```

### LINKED ISSUE #15471: Lag in the editor, replace all hangs Spyder indefinitely
### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [n] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [x] Restarted Spyder
    * [x] Reset preferences with ``spyder --reset``
    * [x] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [x] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
I'm using Python 3.7.10 and Spyder 5.0.1 and I have an i5-4690K CPU on Windows.

I have a very long file, 13k lines. The editor chokes on this and editing code is very laggy and slow. I've turned off as much of the completion/linting as I can, but the issue still occurs. Task Manager shows 1 core + a bit more (~25-33% CPU) is taken up by Spyder's Python instance while the editor lag is occurring.

It's been happening since 4.2.X to the current 5.0.1. The file wasn't as long before 4.2.X so I'm not sure if previous versions have the same issue or not.

I made a tester file that tries to simulate the issues. The problem seems worse in different parts of the code, though I can't really tell what makes it worse.

Here's the tester file: https://www.dropbox.com/s/vzn6ifemrsvv17c/tester.py?dl=0
It's just 13k lines of copied simple nested if-else statements.

And lastly, if you replace anything that occurs a lot with anything else, Spyder will hang and I assume crash eventually. Or it might succeed at some point, but I don't wait around to find out.

Let me know if there's any other info I can provide!

### What steps reproduce the problem?
Warning: Once opening tester.py and replacing "prev" with "\n" made it so Spyder crashes on "Setting up main window". Reset to Spyder allows it to open again. Only happened once.

1. Open tester.py
2. Delete some random lines at the top, go to the bottom and start writing in random if-else or for statements and other things
3. It will slow down heavily (not as bad as my main code which is more complex, sometimes I get ~5 sec pauses)
4. Revert the file and just edit in random if-else and other statements
5. Delete the entire if-else zone and it'll take a minute to work through false line errors
6. Ctrl+R replace "prev" with "\n" and hang the program

### Paste Traceback/Error Below (if applicable)

Just either slow or crashes, no visible errors.

## Versions

* Spyder version:  5.0.1
* Python version:  3.7.10
* Qt version: 5.9.7
* PyQt version: 5.9.2
* Operating System name/version: Windows 10 20H2

### Dependencies

# Mandatory:
atomicwrites >=1.2.0          :  1.4.0 (OK)
chardet >=2.0.0               :  4.0.0 (OK)
cloudpickle >=0.5.0           :  1.6.0 (OK)
cookiecutter >=1.6.0          :  1.7.2 (OK)
diff_match_patch >=20181111   :  20200713 (OK)
intervaltree >=3.0.2          :  3.1.0 (OK)
IPython >=7.6.0               :  7.22.0 (OK)
jedi =0.17.2                  :  0.17.2 (OK)
jsonschema >=3.2.0            :  3.2.0 (OK)
keyring >=17.0.0              :  22.3.0 (OK)
nbconvert >=4.0               :  6.0.7 (OK)
numpydoc >=0.6.0              :  1.1.0 (OK)
paramiko >=2.4.0              :  2.7.2 (OK)
parso =0.7.0                  :  0.7.0 (OK)
pexpect >=4.4.0               :  4.8.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.8.0 (OK)
pygments >=2.0                :  2.8.1 (OK)
pylint >=1.0                  :  2.7.4 (OK)
pyls >=0.36.2;<1.0.0          :  0.36.2 (OK)
pyls_black >=0.4.6            :  0.4.6 (OK)
pyls_spyder >=0.3.2           :  0.3.2 (OK)
qdarkstyle =3.0.2             :  3.0.2 (OK)
qstylizer >=0.1.10            :  0.2.0 (OK)
qtawesome >=1.0.2             :  1.0.2 (OK)
qtconsole >=5.0.3             :  5.0.3 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.8.3                 :  0.9.7 (OK)
setuptools >=39.0.0           :  52.0.0.post20210125 (OK)
sphinx >=0.6.6                :  3.5.3 (OK)
spyder_kernels >=2.0.1;<2.1.0 :  2.0.1 (OK)
textdistance >=4.2.0          :  4.2.1 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog >=0.10.3;<2.0.0      :  1.0.2 (OK)
zmq >=17                      :  20.0.0 (OK)

# Optional:
cython >=0.21                 :  None (NOK)
matplotlib >=2.0.0            :  3.3.4 (OK)
numpy >=1.7                   :  1.20.1 (OK)
pandas >=1.1.1                :  1.2.4 (OK)
scipy >=0.17.0                :  1.6.2 (OK)
sympy >=0.7.3                 :  None (NOK)

### LINKED ISSUE #16975: Replace all is not efficient
### Issue Report Checklist

* [ ] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

### What steps reproduce the problem?
 1. Create a file containing 10000 lines with:
```
with open("test.py", "w") as f:
    for i in range(10000):
        f.write("aaaaaa\n")
```
2. Replace `aaa` by `bbb`
3. Click `All`
4. Spyder freezes for several minutes

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #16975
Fixes #15471
Fixes #14176

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

**Review by ccordoba12 — COMMENTED** [review]
Hey @impact27, thanks for your work on this! And sorry for the delay in reviewing it but it touches some critical code in the editor, so I wanted to review it with care but hadn't have the time to do it.

File: spyder/plugins/editor/utils/decoration.py [hunk 1]
```diff
@@ -37,7 +43,7 @@ class TextDecorationsManager(Manager, QObject):
     def __init__(self, editor):
         super(TextDecorationsManager, self).__init__(editor)
         QObject.__init__(self, None)
-        self._decorations = []
+        self._decorations = {"misc": []}
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/utils/decoration.py:46]: The decorations that end up there have been added by `def add(self, decorations)` without a specific key, so I think `misc` is appropriate here. `current` would imply that the decorations added under a specific key are somehow not current

File: spyder/plugins/editor/widgets/base.py [hunk 2]
```diff
@@ -191,29 +191,17 @@ def set_extra_selections(self, key, extra_selections):
             selection.draw_order = draw_order
             selection.kind = key
 
-        self.clear_extra_selections(key)
         self.extra_selections_dict[key] = extra_selections
-
-    def update_extra_selections(self):
```
[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py:197]: I am not sure how this function would prevent a memory leak? As far as I see it only ever added things in `self.decorations`? But they should be kept up to date and not need to be updated

[the remark this answers is not in the document]
Comment (impact27) [reply spyder/plugins/editor/widgets/base.py:197]: That is the reason why they are not mixed together in an array but rather kept separated by key

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27! I reviewed this carefully and it's working as expected.

Besides, it's a much simpler and straightforward approach to handle decorations. So, thanks a lot for taking the time to improve this part of our code.

**Comment by sphh** [comment]
@impact27 Thanks you so much for this PR! This finally speeds up Spyder!!

I applied those commits manually, made a mistake and got an exception, that the method `update_extra_selections()` is not available (that was my mistake when applying the patches). I corrected this and then I searched for `update_extra_selections` and found an occurrence here (this time not my mistake):
https://github.com/spyder-ide/spyder/blob/6c3b5025e1c9e960c944d6ab2cb4d252f86ae39c/spyder/plugins/editor/widgets/codeeditor.py#L1326
I guess that will throw an exception sooner or later …

**Comment by impact27** [comment]
Thanks!

**Pull request merged**